src/analyzer/protocol/ssl/spicy/SSL.spicy

```diff
@@ -0,0 +1,2058 @@
+    ciphers: uint24[self.ciphers_len / 3];
+    session_id: bytes &size=self.session_len;
+    challenge: bytes &size=self.chal_len;
+
+    on client_version {
+        if (self.client_version != SSLv2 && self.client_version != SSLv3 && self.client_version != TLSv10 && self.client_version != TLSv11 && self.client_version != TLSv12) {
+            spicy::decline_input("Invalid version in SSL client hello. Version: %s, self.client_version"); # Version: " + self.client_version);
+            zeek::skip_input();
+        }
+    }
+
+    on %init {
+        if (msg.first_packet == False) {
+            spicy::decline_input("SSLv2 client hello late in connection");
+        }
+    }
+};
+
+type SSL2ServerHello = unit(len: uint64, msg: Message, inout sh: Share) {
+    direction_check: DirectionCheck(sh, False); # should be sent by responder
+    session_id_hit: uint8 &requires=($$ < 2);
+    certificate_type: uint8 &requires=($$ == 1); # X509 certificate
+    server_version: uint16 &requires=($$ == 2); # Has to be version 2, everything else uses new server hello format
+    cert_len: uint16;
+    ciphers_len: uint16;
+    conn_id_len: uint16;
+    cert_data: bytes &size=self.cert_len;
+    ciphers: uint24[self.ciphers_len / 3];
+    conn_id_data: bytes &size=self.conn_id_len;
+
+    on %init {
+        if (msg.first_packet == False) {
+            spicy::decline_input("SSLv2 server hello late in connection");
+        }
+    }
+
+    on %done {
+        sh.chosen_version_sh_outer = self.server_version;
+        set_version(self.server_version, sh);
+
+        zeek::file_begin("application/x-x509-user-cert", get_certificate_fuid(get_direction(sh), get_and_increase_certificate_depth(sh)));
+        zeek::file_data_in(self.cert_data);
+        zeek::file_end();
+    }
+};
+
+type ClientHello = unit(len: uint64, msg: Message, inout sh: Share) {
+    direction_check: DirectionCheck(sh, True); # should be sent by originator
+    client_version: uint16;
+    random: Random;
+    session_id_length: uint8;
+    session_id: bytes &size=self.session_id_length;
+    dtls_cookie: ClientHelloCookie if(self.client_version == DTLSv10 || self.client_version == DTLSv12);
+    cipher_suites_length: uint16;
+    cipher_suites: uint16[self.cipher_suites_length / 2];
+    compression_methods_length: uint8;
+    compression_methods: uint8[self.compression_methods_length];
+    extensions_length: uint16 if(len > self.offset() + 2);
+    extensions: Extension(sh, True)[] &size=self.extensions_length if(len > self.offset() + 2);
+};
+
+# Draft versions of TLS 1.3 had a different server hello - distinguish here
+type ServerHelloChoice = unit(len: uint64, msg: Message, inout sh: Share) {
+    direction_check: DirectionCheck(sh, False); # should be sent by responder
+    sv: bitfield(16) {
+        server_version0: 0..7;
+        server_version1: 8..15;
+        server_version: 0..15;
+    };
+    var negotiated_version: uint16;
+
+    switch (self.negotiated_version) {
```

Comment:
Is it true that `self.negotiated_version` here is dependent on the `on sv` execution at the bottom? I'm seeing this pattern elsewhere and wonder if this is required for some reason? If not, this appears  obfuscated as reading top to bottom doesn't really work.

Wouldn't it be easier to follow if the hook was directly following the `sv` bitfield?

```
bitfield(16) {
...
} {
    ...
    if (self.sv.server_version1 == 0x7F) {
        self.negotiated_version = 0x7f00;
    ...
}

switch (self.negotiated_version) {
}
```

The pattern of having `on field { }` at the end of the unit is used pretty often, so wonder generally if there's a reason rather than following them directly with the parsed field.

Reply:
I think I agree with @awelzel's sentiment here, having "smallish" hooks at the end makes the flow harder to follow for no good reason. Being able to read a parser top-down is useful and makes side effects easier to understand.

In this case the hook directly influences parsing of following fields and IMO hiding that away in a hook "elsewhere" makes this less transparent; since the code is small enough I'd tend to just declare it inline. (potentially different story if we e.g., fill up some unit var only used in an EVT hook). This is also relevant when thinking about error recovery where being able to quickly see which side effects parsing a field has makes it easier to reason about the state of fields we see; IMO an out-of-line field hook has a tendency to be mistaken for a unit `%done` hook.

Overall I think when possible it is worth taking advantage of Spicy's ability to mix declarative (fields) and procedural code.